### DISCUSSION TITLE
Null safety with intellij inspections

### DISCUSSION THREAD

**Discussion by robp94** [mapstruct/mapstruct#2643] [Ideas]
I think we found a way to have some kind of null safety with map struct. If you have two objects like this:

```java
    public class Object1{
        @NotNull
        String test = "";
    }

    public class Object2{
        @Nullable
        String test = null;
    }
```
and do a mapping like this;
```java
        var object1 = new Object1();
        var object2 = new Object2();

        object1.test = object2.test; <- Intellij warns you that this could be null
```
Intellij, teamcity or sonarqube (probably there are other tools as well) can warn you or even break your build. For the simplest mappings, it would be enough to only remove the `@Generated` annotation in the class and mark the folder as not generated. This is currently not possible to configure. Could this be added?

For more complicated mappings like with multiple parameters, we would need some more control, but maybe they would be not too complicated to add.

For example:
```java
    public Tool toTool(ToolData toolData, UUID id) {
        if ( toolData == null && id == null ) {
            return null;
        }

        String name = null;
        Address address = null;
        if ( toolData != null ) {
            name = toolData.getName();
            address = toolData.getAddress();
        }
        UUID id1 = null;
        id1 = id;

        Tool tool = new Tool( id1, name, address );

        return tool;
    }
```
If we have code like this, we also get a warning that name and address could be null if even we would make sure that both parameters are never null. This could be solved with this https://github.com/mapstruct/mapstruct/discussions/2642 or if we would have an option to remove the null checks for the parameters. 

Another feature, which would be useful for this, would be if annotations on the parameters would get added to the generated code. Like this:
interface
```java
Tool toTool(@NotNull ToolData toolData, @NotNull UUID id)
```
and then have it in the generated code above.

**Comment by filiphr** [comment]
> For the simplest mappings, it would be enough to only remove the `@Generated` annotation in the class and mark the folder as not generated. This is currently not possible to configure. Could this be added?

I am not sure what you are asking here. You want to configure to not add `@Generated` or do a null value check always? For the null value check we have the `NullValueCheckStrategy` that can be put on the `@BeanMapping`, `@Mapper` or `@MapperConfig`. For the `@Generated` we have a compiler option that can disable.

We have an open feature request #1243 to add support for the `Nullable` / `NotNull` annotations. If we do that then we would be able to generate code that would pass the validations from the other tools.

By the way even if we pass the annotations down to the generated method (which is a good and separate thing anyways) the tools will still complain for the 

```
        if ( toolData == null && id == null ) {
            return null;
        }
```

because a null check is done for something that is not null. That takes us back to #1243.

**Reply by robp94** [reply]
Yeah it would only be some workaround until we have proper support. We use Kotlin, so if you try to set a nullable value into a nonnull value, you get a compile error. I think in Java you would although not want to set a nullable value into a nonnull value, even if it is possible it is probably not wanted. Currently, with mapstruct there is no safety mechanism to warn about such problems/errors, for us it would be nice to have some. 

I think the `NullValueCheckStrategy` does not help in this situation. We would want to get notified that the schema is maybe faulty, or we need some kind of logic to provide a default value or something like this.

> For the `@Generated` we have a compiler option that can disable.

In the `@Generated` annotation, there are 3 values (value, date, comments), I only found options too disable two of them. `mapstruct. suppressGeneratorTimestamp` and `mapstruct. suppressGeneratorVersionInfoComment`, so if you disable both the annotation is not generated or did I overlook something?

> the tools will still complain for the

I know I mentioned it above that for this there would be some more control needed. But maybe it could be done with little effort.

> If we have code like this, we also get a warning that name and address could be null if even we would make sure that both parameters are never null. This could be solved with this #2642 or if we would have an option to remove the null checks for the parameters.
